CONTRIBUTING.md

```diff
@@ -0,0 +1,57 @@
+
+Here are a few tags Coq developers may add to your PR and what they mean. In general feedback and requests for you as the pull request author will be in the comments and tags are only used to organize pull requests.
+
+- [needs: rebase](https://github.com/coq/coq/pulls?utf8=%E2%9C%93&q=is%3Aopen%20is%3Apr%20label%3A%22needs%3A%20rebase%22%20) indicates the PR should be rebased on top of the latest `master` branch. See the [GitHub documentation](https://help.github.com/articles/about-git-rebase/) for a brief introduction to using `git rebase`.
+- [needs: fixing](https://github.com/coq/coq/pulls?q=is%3Aopen+is%3Apr+label%3A%22needs%3A+fixing%22) indicates the PR needs a fix, as discussed in the comments.
+- [needs: testing](https://github.com/coq/coq/pulls?q=is%3Aopen+is%3Apr+label%3A%22needs%3A+testing%22) indicates the PR needs testing. This is often used when testing beyond what the test suite can handle is required. For example, performance benchmarking is currently performed with a different infrastructure. Unless some followup is specifically requested you aren't expected to do this additional testing.
+
+## Documentation
+
+Currently the process for contributing to the documentation is the same as for changing anything else in Coq, so please submit a pull request as described above.
+
+Bugzilla includes a component to mark bugs related to documentation. You can view a list of documentation-related bugs using a [Bugzilla search](https://coq.inria.fr/bugs/buglist.cgi?component=Doc&list_id=455006&product=Coq&resolution=---). Many of these bugs can be fixed by contributing writing, without knowledge of Coq's OCaml source code.
+
+The sources for the [Coq reference manual](https://coq.inria.fr/distrib/current/refman/) are at [`doc/refman`](/doc/refman). These are written in LaTeX and compiled to HTML with [HeVeA](http://hevea.inria.fr/).
+
+## Contributing outside this repository
+
+There are many useful ways to contribute to the Coq ecosystem that don't involve the Coq repository.
+
+Tutorials to teach Coq, and especially to teach particular advanced features, would be much appreciated. Some tutorials are listed on the [Coq website](https://coq.inria.fr/documentation). If you would like to add a link to this list, please make a pull request against the Coq website repository at https://github.com/coq/www.
+
+Subscribe to the [coq-club](https://coq.inria.fr/community) mailing list and answer questions there. Asking your own questions on the mailing list will also help others learn from the answers, which often come from experts in the community.
+
+Ask and answer questions on [Stack Exchange](https://stackexchange.com/filters/299857/questions-tagged-coq-on-stackexchange-sites) which has a helpful albeit small community of users.
```

Comment:
At DSSS17 there was a bunch of discussion about what would be the best forum for asking and answering questions, and the conclusion was that right now it is indeed Stack Exchange.  So maybe the advice (that people wishing to be helpful should hang out there and answer -- and ask! -- questions) should be strengthened (and repeated elsewhere, like the main web page).  The isabelle community made a decision a little while ago to use StackExchange this way, and I'm told it has worked out very well for them.

Reply:
@bcpierce00 Can you summarize the reasons behind the conclusion at DSSS17 to prefer Stack Exchange?
(Personally, I find the culture to be quite toxic there.)

Reply:
@aa755 
> Personally, I find the culture there to be very toxic.

Can you elaborate? I am genuinely interested to know what in the culture you find toxic (what kind of behaviors, patterns...). As you can see [in this blog post](http://www.theozimmermann.net/2016/12/stackoverflow/), I've been exploring / getting to understand Stack Exchange quite a lot recently.

Reply:
The factors I remember are:
- Lightweight, easy for people to contribute, good indexing (by google)
- Alternative technologies (in particular, wikis) don't work as well -- too hard to keep from bitrotting.
- Isabelle community has been happy there

I've also heard some negative things about the culture there (e.g., strange rules about what count as good questions), which led to some discussion about whether one should propose a separate Coq Exchange on the SE platform.  But that is a longer-term question.

Reply:
@Zimmi48: My main problem with stackexchange is that lately the moderators have too much power and/or they are frequently abusing it to close many good questions for incorrect reasons. 
(Diamond moderators can unilaterally close a question, even against the votes of other moderators.)
Also, there are many people there who create nuisance just to get reputation: e.g. making minor, sometimes semantics-altering edits.
As a result, I don't feel good about contributing there. It seems I am not the only one with such sentiments: https://hackernoon.com/the-decline-of-stack-overflow-7cb69faa575d

IMHO, Coq-club is almost as good as stack exchange on most factors that Benjamin mentioned (Thanks Benjamin). 
I would argue that coq-club is more lighweight, especially for people who are new to stackexchange: they don't have to get reputation before they can fully participate.
AFAIK, coq-club is also indexed by Google. If not, I will be happy to help in setting up the indexing. Also, AFAIK, one can download all the discussions of coq-club from GMane and search the archives in whatever way they like. Even if GMane stops working, I guess Inria would be happy to provide a way to download all archives.

In summary, I think that deprecating coq-club (e.g. by removing its mentions) is a bad idea because coq-club may be a more welcoming place for many legitimate questions.
I'm fine with mentioning both coq-club and stack-exchange.

Reply:
At least some people have the impression that beginner questions (and beginners in general) are not that welcome on coq-club...  

This may be a misimpression, but even so: the fact that it is an email venue -- i.e., asking a question is going to cause some unknown number of people to drop whatever they are doing and switch to reading my email -- is certainly a deterrent for _me_ to ask random questions on coq-club.  :-)

Reply:
@aa775 does this happen across all languages / topics, or only on popular ones?  My experience with math.se and tex.se has been fairly different: there is a relatively small set of true experts on any given topic and they answer most of the questions that show up there.  I'd expect a similar thing for, e.g., Haskell questions on SE, and I'm curious to know if this is the case.  I'd predict that as long as the set of people who are proficient at Coq remains small (where small means "not much more than TeX or math PhDs or Haskell"), we won't run into the toxic community issues on SE.

Reply:
@bcpierce00 I think my impression is that beginner questions on coq-club tend to get "read this chapter of CPDT", while beginner questions on IRC get you a back and forth on what issues you're working on, and this might make irc more beginner friendly?

Reply:
@aa755 
> lately, the moderators have too much power and/or they are frequently abusing it to close many good questions for incorrect reasons

I've never remarked such a problem, but this might be due to the fact that I'm only following "niche" tags such as Coq, Elm, Nix and the whole (small) Open Source SE. It might be a very different experience with Java, Python, Javascript...

> I would argue that coq-club is more lighweight, especially for people who are new to stackexchange: they don't have to get reputation before they can fully participate.

I agree. That being said, as I'm sort of explaining in my blog post, the goal of Stack Exchange is not to have everyone participate but rather to get a good set of canonical questions to which later users can refer too (and find easily by Googling). Answers stay more up-to-date because people are encouraged to edit them whenever the solution evolves, which is not the case of a mailing list. IMO mailing lists are also much harder to browse, even when they are indexed.

> In summary, I think that deprecating coq-club (e.g. by removing its mentions) is a bad idea.

We didn't go as far as deprecating it / removing its mentions fortunately. This contributing guide did not exist before and the argument for it to not mention Coq-Club was that Coq-Club has already a large mention in the README. Similarly, I'm proposing to add Stack Exchange on the community page of the website, not to remove the existing mention of Coq-Club.

@bcpierce00 
> I've also heard some negative things about the culture there (e.g., strange rules about what count as good questions), which led to some discussion about whether one should propose a separate Coq Exchange on the SE platform.

That is a question which would be worth discussing. For now, most Coq questions end up on StackOverflow, which is the platform for "programming" question. And although Coq *is* a programming language, I don't know how appropriate this is. For now, AFAIK, no one external to the Coq community has bothered us with this, in particular there have been no intervention of moderators (again AFAIK). An alternative, equally appropriate, venue for Coq would be CS Stack Exchange. It seems to me that Coq questions are equally welcome there. Finally there is CS theory SE for questions on the theory behind Coq, and Code Review SE which could be used more for showing one's proof and getting advice on how to do it differently. That Coq questions can be spread across multiple sites is why I provided a link that allows to browse questions tagged Coq across all sites.

Last remark, our community being so small, we have few people with high reputation and thus moderator privileges. It is important to up-vote each other' answers to give more power to the members of this community.

Reply:
I tend to prefer mailing lists; maybe a `coq-beginners` mailing list could be useful, similarly, a #beginners channel on Gitter.

Reply:
@JasonGross @Zimmi48 :

SE moderators do sometimes impose their authority on niche topics that they have no clue about.
Here are some relevant examples:
https://meta.stackexchange.com/questions/180900/question-on-isabelle-proof-by-cases-shouldnt-have-been-closed
https://stackoverflow.com/questions/19421234/how-do-i-generate-latex-from-isabelle-hol
https://stackoverflow.com/questions/23084610/had-a-verified-ssl-tls-implementation-in-a-safe-language-would-it-still-have-b
https://stackoverflow.com/questions/40848265/isabelle-and-scala
https://stackoverflow.com/questions/31306654/is-there-a-translator-from-haskell-to-coq

Also, consider the fact that there may be many other legit questions that just got deleted because they were closed before they got any answer. 
Personally, I have encountered too-authoritarian moderators at other SE sites too:
https://physics.stackexchange.com/questions/355267/viewing-totality-though-an-unfiltered-telescope

Reply:
I do like SE. I'm generally happy with the moderation; I find it pretty good. Sometimes mistakes are made. Thanks for the links @aa755: I voted to reopen the questions I thought  appropriate for SE.

Reply:
This discussion shows that people may have various preferences and we should avoid to have only one channel for asking questions. Coq-Club is already well known; I am happy that we are now also mentioning Stack Exchange; maybe we should have a Gitter beginners channel as Emilio suggested: it is true that beginners' questions may not always have their place on Coq-Club nor Stack Exchange, especially when they "do not show sufficient research". Elm's Slack beginners channel is pretty successful I think (but let's not use Slack, it is being abused as a community discussion platform while it was never created to be one and limits the size of the archives you can browse).

Reply:
I have some concerns about fragmentation and persistence of knowledge related to Coq.
I don't have any solutions, though.

# Fragmentation:
When I have a question of Coq, I would like to have one place to find the answers. Google is okay for the first attempt. But I often benefit by going beyond Google: I often also search the archives of coq-club (emails after 2013 are in my inbox. I downloaded the rest from GMane).
If we have too many places that we encourage asking questions, and no good way to *exhaustively* search those places I worry we may effectively loose some valuable knowledge.

# Persistence
Questions asked on Coq-club are guaranteed to live forever. I don't know about any such guarantees about SE or other platforms managed by for-profit companies.

Reply:
Have you heard about SE dumps? SE lets you download their full archives in a convenient format, so I'm not too worried about persistence: https://archive.org/details/stackexchange .

One thing I'm concerned about is obsolescence: I keep running into obsolete info on Coq-club.

Reply:
Personally, I often find obsolete answers useful, as long they don't prevent me from finding the non-obsolete answer (so far, they never prevented me: modern tools allow things like sorting by date). Obsolete answers give me a useful historical perspective on how Coq evolved. Just yesterday, in Coq-club archives, I found it interesting to read 11-year-old answers from experts like Benjamin Werner, even though I knew those answers were obsolete.

Reply:
On these two aspects, SE is much more performant than Coq-Club or any other mailing list IMHO.
You, Abhishek, are a very experimented Coq user and you have efficient way of searching for useful information but to a novice or even to a less experimented Coq user, it would be much more profitable if SE contained more Coq questions / answers because these are so easy to find thanks to Google. It is *not* the case of Coq-Club.
About persistence, I'll add that all the content on SE is licensed under CC-BY-SA which allows for re-uploading elsewhere, compilation, derivation...